tests/test_simulation.py

```diff
@@ -329,3 +494,212 @@ def test_sequence_sample_dnsm(pcp_df, dnsm_burrito):
+            aa_mask = _token_aa_mask[: len(aa_parent)]
+            # sel_matrix = torch.ones((aa_seq_len, 20))
+            # This is in linear space.
+            sel_matrix = generic_burrito.build_selection_matrix_from_parent_aa(
+                _token_aa_parent_idxs, _token_aa_mask
+            )[: len(aa_parent)]
+            # neutral_sel_matrix[torch.arange(aa_seq_len), aa_parent_idxs]
+
+            # First way:
+            nt_mut_probs = 1.0 - torch.exp(-branch_length * nt_rates)
+            codon_mutsel, _ = molevol.build_codon_mutsel(
+                parent_idxs.reshape(-1, 3),
+                nt_mut_probs.reshape(-1, 3),  # Linear space
+                nt_csps.reshape(-1, 3, 4),  # Linear space
+                sel_matrix,  # Linear space
+                multihit_model=multihit_model,
+            )
+            log_codon_mutsel = clamp_probability(codon_mutsel).log()
+            flat_log_codon_mutsel = molevol.flatten_codons(log_codon_mutsel)
+            flat_log_codon_mutsel[~aa_mask] = float("nan")
+
+            # Second way:
+            neutral_codon_probs = molevol.neutral_codon_probs_of_seq(
+                seq,
+                aa_mask,
+                nt_rates,
+                nt_csps,
+                branch_length,
+                multihit_model=multihit_model,
+            )
+            adjusted_codon_probs = molevol.adjust_codon_probs_by_aa_selection_factors(
+                codon_parent_idxs.unsqueeze(0),
+                neutral_codon_probs.unsqueeze(0).log(),
+                sel_matrix.unsqueeze(0).log(),
+            ).squeeze(0)
+
+            # Now let's compare to the simulation probs:
+            sim_probs = clamp_probability(
+                codon_probs_of_parent_seq(
+                    generic_burrito.to_crepe(),
+                    (seq, ""),
+                    branch_length,
+                    neutral_crepe=neutral_crepe,
+                    multihit_model=multihit_model,
+                )[0]
+            ).log()
+
+            # These atol values can made more strict if netam.common.SMALL_PROB
+            # is made smaller, so the differences are just cause of different
+            # amounts of clamping in different code paths
+
+            # Compare mutsel path with adjust by selection factors path:
+            if not torch.allclose(
+                adjusted_codon_probs, flat_log_codon_mutsel, equal_nan=True, atol=1e-07
+            ):
+                diff_mask = ~torch.isclose(
+                    adjusted_codon_probs, flat_log_codon_mutsel, equal_nan=True
+                )
+                print(flat_hit_classes[diff_mask])
+                print((adjusted_codon_probs - flat_log_codon_mutsel)[diff_mask])
+                print(adjusted_codon_probs[diff_mask])
+                print(flat_log_codon_mutsel[diff_mask])
+                assert False
+
+            # adjusted_codon_probs = molevol.zero_stop_codon_probs(clamp_probability(adjusted_codon_probs.exp()).log())
```

Comment:
I have not taken a close read of these functions-- I trust that they are doing what you want-- but perhaps you want to make a quick scan to tidy things up. Is this useful or cruft?